In [1]:
import sklearn as sk
import pandas as pd

## Preprocesamiento

In [2]:
equipos = pd.read_csv('../Data/equipos_data.csv')
equipos = equipos[['equipo','posicion','valor_mercado','aforo']]
historico = pd.read_csv('../Data/historico_data.csv')
historico = historico[['equipo1','equipo2','mes','fin_de_semana','publico']]

##### Ver si se llenan los estadios en algunos partidos 

In [3]:
se_llena = pd.merge(historico,equipos, left_on='equipo1', right_on='equipo')
se_llena[se_llena['publico'] == se_llena['aforo']]

equipo1         equipo2  mes  fin_de_semana  publico  \
5     Real Madrid CF    FC Barcelona    3              1    81044   
8     Real Madrid CF    FC Barcelona    2              1    81044   
21    Real Madrid CF      Sevilla FC    4              1    81044   
49    Real Madrid CF   Real Sociedad   11              1    81044   
54    Real Madrid CF   Villarreal CF    8              1    81044   
...              ...             ...  ...            ...      ...   
2693       Getafe CF      Sevilla FC    2              1    16800   
2751       Getafe CF      CA Osasuna    3              1    16800   
2887    RCD Mallorca      Sevilla FC    4              1    20500   
3043        Cádiz CF  Real Madrid CF    2              1    21094   
3394      UD Almería        Cádiz CF    4              1    15274   

              equipo  posicion  valor_mercado  aforo  
5     Real Madrid CF         1         770.75  81044  
8     Real Madrid CF         1         770.75  81044  
21    Real Madrid CF         1         770.75  81044  
49    Real Madrid CF         1         770.75  81044  
54    Real Madrid CF         1         770.75  81044  
...              ...       ...            ...    ...  
2693       Getafe CF        15         208.68  16800  
2751       Getafe CF        15         208.68  16800  
2887    RCD Mallorca        16          74.15  20500  
3043        Cádiz CF        17          67.80  21094  
3394      UD Almería        18          63.75  15274  

[157 rows x 9 columns]

Se llena el estadio en 157 partidos

### Obtener variables dummies para entrenar los modelos

In [4]:
historico = pd.get_dummies(historico, prefix=['equipo1','equipo2'])

## Modelos Predictivos

In [5]:
X = historico.drop('publico',axis=1)
y = historico['publico']

### Regresion Lineal

In [6]:
from sklearn.linear_model import LinearRegression
# from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error
clf = LinearRegression(normalize=True)
clf.fit(X,y)
scores = clf.predict(X)
mean_squared_error(y,scores)**(1/2)

c:\Users\migue\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:145: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), LinearRegression())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)


  FutureWarning,


4911.61264786135

### Random Forest

In [7]:
from sklearn.ensemble import RandomForestRegressor
forest = RandomForestRegressor(max_depth = 100,random_state=0)
forest.fit(X, y)
scores = forest.predict(X)
mean_squared_error(y,scores)**(1/2)

2488.684547167605

In [26]:
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsRegressor

knr = KNeighborsRegressor(n_neighbors=5)
knr.fit(X, y)
mean_squared_error(y, knr.predict(X))**0.5

8456.288391589826